beacon-chain/db/filesystem/data_column.go

```diff
@@ -0,0 +1,966 @@
+// This function expects all data column sidecars to belong to the same block.
+func (dcs *DataColumnStorage) saveFilesystem(root [fieldparams.RootLength]byte, epoch primitives.Epoch, dataColumnSidecars []blocks.VerifiedRODataColumn) error {
+	// Compute the file path.
+	filePath := filePath(root, epoch)
+	exists, err := afero.Exists(dcs.fs, filePath)
+	if err != nil {
+		return errors.Wrap(err, "afero exists")
+	}
+
+	dcs.pruneMu.RLock()
+	defer dcs.pruneMu.RUnlock()
+
+	fileMu, toStore := dcs.fileMutex(root)
+	toStore <- dataColumnSidecars
+
+	fileMu.Lock()
+	defer fileMu.Unlock()
+
+	if exists {
+		if err := dcs.saveDataColumnSidecarsExistingFile(filePath, toStore); err != nil {
+			return errors.Wrap(err, "save data column existing file")
+		}
+
+		return nil
+	}
+
+	if err := dcs.saveDataColumnSidecarsNewFile(filePath, toStore); err != nil {
+		return errors.Wrap(err, "save data columns new file")
+	}
+
+	return nil
+}
+
+// Get retrieves data column sidecars from the database.
+// If one of the requested data column sidecars is not found, it is just skipped.
+// If indices is nil, then all stored data column sidecars are returned.
+// Since DataColumnStorage only writes data columns that have undergone full verification, the return
+// value is always a VerifiedRODataColumn.
+func (dcs *DataColumnStorage) Get(root [fieldparams.RootLength]byte, indices []uint64) ([]blocks.VerifiedRODataColumn, error) {
+	dcs.pruneMu.RLock()
+	defer dcs.pruneMu.RUnlock()
+
+	fileMu, _ := dcs.fileMutex(root)
+	fileMu.RLock()
+	defer fileMu.RUnlock()
+
+	startTime := time.Now()
+
+	// Build all indices if none are provided.
+	if indices == nil {
+		indices = make([]uint64, mandatoryNumberOfColumns)
+		for i := range indices {
+			indices[i] = uint64(i)
+		}
+	}
+
+	// Preventive check of indices.
+	for _, index := range indices {
+		if index >= mandatoryNumberOfColumns {
+			return nil, ErrDataColumnIndexTooLarge
+		}
+	}
+
+	summary, ok := dcs.cache.get(root)
+	if !ok {
+		// Nothing found in db. Exit early.
+		return nil, nil
+	}
+
+	// Compute the file path.
+	filePath := filePath(root, summary.epoch)
+
+	// Open the data column sidecars file.
+	// We do not specially check if the file exists since we have already checked the cache.
+	file, err := dcs.fs.Open(filePath)
+	if err != nil {
+		return nil, errors.Wrap(err, "data column sidecars file path open")
+	}
+
+	// Read file metadata.
+	metadata, err := dcs.metadata(file)
+	if err != nil {
+		return nil, errors.Wrap(err, "metadata")
+	}
+
+	// Retrieve data column sidecars from the file.
+	verifiedRODataColumnSidecars := make([]blocks.VerifiedRODataColumn, 0, len(indices))
+	for _, index := range indices {
+		// Skip if the data column is not saved.
+		if metadata.indices[index] < limit {
+			continue
+		}
+
+		// Compute the offset of the data column sidecar.
+		offset := headerSize + int64(metadata.indices[index]-limit)*int64(metadata.sszEncodedDataColumnSidecarSize)
```

Comment:
While having the zero value mean unset is intuitive and is clearly safer, I keep thinking if all the unset fields had their high bit set when the metadata section is initialized, you never have to do conversions on the offsets, so it actually simplifies the code a little. eg if the value < 128 it would be set, >= 128, unset.

Reply:
I'm not either! And to be clear it isn't a question of "efficiency" or something, more about remembering that when you interact with this byte value you need to add/subtract.

I just had an idea I like though, what about encapsulating it in a type like this?

```
type StorageIndex byte

func (i StorageIndex) IsSet() bool

func (i StorageIndex) Position() int

func (i StorageIndex) Set(pos int)

```

Reply:
nice!